File: solution-cpp/022_generate_parentheses.cpp
/*
	022. Generate Parentheses
	@name: li jin
	@date: Mar 03, 2016
	@link: https://leetcode.com/problems/generate-parentheses/
	@time: 0 ms
*/

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> strs;
        genParis(strs, "", n, 0);
        return strs;
    }

    void genParis(vector<string> &strs, string str, int left, int right) {
        if (left == 0 && right == 0) {
            strs.push_back(str);
            return;
        }
        if (left > 0) { genParis(strs, str + '(', left - 1, right + 1); }
        if (right > 0) { genParis(strs, str + ')', left, right - 1); }
    }
};

File: solution-python/022_generate_parentheses.py
"""
	022. Generate Parentheses
	@name: li jin
	@date: Mar 03, 2016
	@link: https://leetcode.com/problems/generate-parentheses/
	@time: 48 ms
"""

class Solution(object):
    def generateParenthesis(self, n):
        ans = []
        def generate(s, ln, rn):
            if ln == 0 and rn == 0:
                ans.append(s)
            else:
                if ln:
                    generate(s + '(', ln-1, rn+1)
                if rn:
                    generate(s + ')', ln, rn-1)
        generate("", n, 0)
        return ans;
